src/parser/ASTAllocator.cpp

```diff
@@ -43,20 +43,27 @@ void ASTAllocator::reset()
     ASSERT(m_astPoolMemory != nullptr && m_astPoolEnd != nullptr);
     ASSERT(static_cast<size_t>(m_astPoolEnd - m_astPoolMemory) >= 0);
 
-    for (size_t i = 0; i < m_astPools.size(); i++) {
-        free(m_astPools[i]);
-    }
-    m_astPools.clear();
+    if (m_astPools.size()) {
+        m_astPoolMemory = static_cast<char*>(m_astPools[0]);
 
-    m_astPoolMemory = static_cast<char*>(currentPool());
+        free(m_astPoolEnd - astPoolSize());
+        for (size_t i = 1; i < m_astPools.size(); i++) {
+            free(m_astPools[i]);
+        }
+        m_astPools.clear();
+        m_astPoolEnd = m_astPoolMemory + astPoolSize();
+    } else {
+        m_astPoolMemory = static_cast<char*>(currentPool());
+        m_astPoolEnd = m_astPoolMemory + astPoolSize();
```

Comment:
This line can be removed because current pool is reused without any change (`m_astPoolEnd` has no change).

Reply:
You are right!